### PR TITLE
This module simplifies acknowledging contributions made through GitHub.

### PR DESCRIPTION
**Pull request by tobie**
Fetches names of contributors from github and uses them to fill
in the content of elements with key identifiers:
`#gh-commenters`: people having contributed comments to issues.
`#gh-contributors`: people whose PR have been merged.
Spec editors get filtered out automatically.

/cc @marcoscaceres

### CODE REVIEW

File: js/core/contrib.js [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+                        return github.fetch(url);
+                    })).then(function () {
+                        var names = slice(arguments).map(function (user) {
+                            user = user[0];
+                            return user.name || user.login;
+                        }).filter(function (name) {
+                            return editors.indexOf(name) < 0;
+                        });
+                        names.sort(function (a, b) {
+                            return a.toLowerCase().localeCompare(b.toLowerCase());
```
Comment (marcoscaceres) [inline js/core/contrib.js]: nice touch with localeCompare :)

File: js/core/contrib.js [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+                        });
+                        $(element).html(join(names)).attr("id", null);
+                    });
+                }
+
+                msg.pub("start", "core/contrib");
+                var $commenters = $(doc).find("#gh-commenters")[0];
```
Comment (marcoscaceres) [inline js/core/contrib.js]: Maybe we should stop relying on jQuery here? a simple querySelector would do.

File: js/core/contrib.js [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+                var $contributors = $(doc).find("#gh-contributors")[0];
```
Comment (marcoscaceres) [inline js/core/contrib.js]: as here.

File: js/core/contrib.js (hunk 1 above)
Comment (tobie) [inline js/core/contrib.js]: MDN copy and paste. You should know my coding style by now.

File: js/core/contrib.js [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+                if (conf.githubAPI && ($commenters || $contributors)) {
+                    github.fetch(conf.githubAPI).then(function (json) {
+                        return $.when(
+                            github.fetchIndex(json.issues_url),
+                            github.fetchIndex(json.issue_comment_url),
+                            github.fetchIndex(json.contributors_url)
+                        ).then(function(issues, comments, contributors) {
```
Comment (marcoscaceres) [inline js/core/contrib.js]: Whoa! The API gives you back more than one single value in the returned promise????

File: js/core/contrib.js [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+                            var editors = respecConfig.editors.map(prop("name"));
+                            var commenters = findUsers(issues, comments);
+                            contributors = contributors.map(prop("url"));
+                            return $.when(
+                                toHTML(commenters, editors, $commenters),
+                                toHTML(contributors, editors, $contributors)
+                            );
+                        });
+                    }).then(theEnd);
```
Comment (marcoscaceres) [inline js/core/contrib.js]: no .catch() needed?

Comment (marcoscaceres) [inline js/core/contrib.js]: You should 
`return;`
here. and remove the `} else {` and the just call `theEnd();`

File: js/core/contrib.js [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+                    };
+                }
+                
+                function slice(args) {
+                    return Array.prototype.slice.call(args, 0)
+                }
+
+                function findUsers() {
+                    var users = {};
+                    slice(arguments).forEach(function (things) {
```
Comment (marcoscaceres) [inline js/core/contrib.js]: Nit: this nested .forEach bothers me, tbh. These kinds of functions are supposed to be chained, not nested. It probably means that the first forEach is not what you want here?

**Comment by marcoscaceres** [comment]
Generally, this looks really good @tobie! just a couple of nits.

File: js/core/contrib.js (hunk 4 above)
Comment (tobie) [inline js/core/contrib.js]: JQuery promises are all kinds of weird.

File: js/core/contrib.js (hunk 2 above)
Comment (tobie) [inline js/core/contrib.js]: Indeed.

File: js/core/contrib.js (hunk 6 above)
Comment (tobie) [inline js/core/contrib.js]: Ideally you'd need `flatten()` here, followed by `uniq`. But JS doesn't have either, so I'm not sure what else to do here.

File: js/core/contrib.js (hunk 5 above)
Comment (tobie) [inline js/core/contrib.js]: Yeah I got lazy and forgot how to report errors in ReSpec. Should do.

**Comment by darobin** [comment]
+1

**Comment by tobie** [comment]
@marcoscaceres: I don't know how to address or what to do about your concerns re nested `forEach` loops, nor, tbh, what those concerns precisely are ;). Other than that, this is ready to merge as far as I'm concerned.

**Comment by darobin** [comment]
I don't understand the `forEach` thing either — if you gentlemen agree on the nitpicks please go ahead and merge!

**Comment by marcoscaceres** [comment]
:ship: IT!!!!

**Comment by marcoscaceres** [comment]
> I don't know how to address or what to do about your concerns re nested forEach loops, nor, tbh, what those concerns precisely are ;)

The concern is just nested loops are ugly and fragile: consider, if we wanted to then `.map` or `.filter` inside there later we would end up with a confusing pyramid of nested scopes. That's why you really want atomic functional operations: 
```
let result = list
   .map(item => ...)
   .filter(item => ...)
   .map(item => ...)
   .reduce(item => ...)
```

Rather than (contrived example and not a reflection of your code!!!, but you get the idea):

```JS
list.forEach( thing => {
   things.forEach( thing => {
      thing.map( 
          thing => filter()
      })   
  })
)
```
It's a nit, tho - so no worry.

**Comment by tobie** [comment]
Heh, I know you're not out to get at my coding style, so don't worry. I'm just genuinely trying to figure out how you'd write this differently in a way that doesn't suck.

Here's what it would look like using only the functional array tools available:

```js
function findUsers() {
    return slice(arguments).reduce(function(output, thing) {
        output.push.apply(output, thing);
        return output;
    }, []).filter(function (thing) {
        return thing.user;
    }).map(function(thing) {
        return thing.user.url;
    }).reduce(function(output, url) {
        if (output.indexOf(url) < 0) output.push(url);
        return output;
    }, []);
}  
```

I tend to find that style a slightly dreadful use of `reduce` but YMMW.

Of course, in PrototypeJS that would have been written as:

```js
function findUsers() {
    return slice(arguments).flatten().pluck("user").compact().pluck("url").uniq();
}  
```

…which is nice, sweet and expressive.

LoDash isn't bad either:

```js
function findUsers() {
    return _(arguments).flatten().pluck("user").compact().pluck("url").uniq().value();
} 
```

I can't believe these things aren't in ES proper at this point.

**Comment by tobie** [comment]
ES6 makes the first example look a little less bad:

```js
function findUsers() {
    return slice(arguments)
        .reduce((output, x) => {
            output.push.apply(output, x);
            return output;
        }, [])
        .filter(x => x.user)
        .map(x => x.user.url)
        .reduce((output, url) => {
            if (output.indexOf(url) < 0) output.push(url);
            return output;
        , []);
}
```

But still. Those `.reduce()` calls… ew!

**Comment by marcoscaceres** [comment]
Argh, yeah... so depressing not having `.pluck()` and friends in ES. You know... we have JQuery already, why not include LoDash? Makes a lot of stuff way more legible.

**Comment by darobin** [comment]
I'm not sure that lodash would gain us that much. I think that switching to Browserify+ES6 using Babel would get us a lot of clarity (and possibly the ability to ditch HBS), I'd focus more on that.

**Comment by tobie** [comment]
HBS?

**Comment by darobin** [comment]
It's shipped :rocket:
HBS=Handlebars

**Pull request merged**